README
NOTE
====
This is the last version of hddtemp, 0.3beta-15, with every Debian patch up to Sid
from https://packages.debian.org/sid/hddtemp (-53)

New features
============
* control the cache interval for daemon down to 2 seconds from default of 60

DESCRIPTION
===========
hddtemp will show you the hard drive temperature if the hard drive can
(modern drives support this feature).


COPYRIGHT
=========
This software is distributed under the GPL license.
Please see COPYING file.


INSTALLATION
============
Run ./configure (see --help for options), then just type 'make' and don't forget
to get the last hddtemp.db file on the site :

$ wget http://www.guzu.net/linux/hddtemp.db

By default, the hddtemp.db will we searched in /usr/share/misc directory but
you can change this at compilation time with the --with-db-path option:

$ ./configure --with-db-path=/etc/hddtemp.db



INFORMATION
===========
hddtemp accesses to the SATA disks via ATA pass-through commands (defined in 
T10/04-262r7). Only kernel >= 2.6.16 have this support.

In daemon mode, hddtemp doesn't allow too much query at a time. If the interval
between two query is inferior to 1 minute, hddtemp will give the previous value
(hddtemp query the disk for each connection, if the period is superior to 1
minute).
After starting hddtemp in daemon mode, you can test it with a simple telnet
or netcat:

$ telnet localhost 7634
or
$ netcat localhost 7634

(note: sometimes the netcat command is called nc).

If you know that your drive has a temperature sensor and is reported to be
unsupported tell me which model and which manufacturer it is, and/or just
add a new entry in hddtemp.db. Each line of hddtemp.db is either a commentary,
a blank line or a line containing:
 - a regular expression that allow hddtemp to recognize a drive or a set of drives
   from its model name or from a generic model name,
 - a value (ATTRIBUTE_ID from S.M.A.R.T.),
 - a description.

The separator option allows to specify a separator in the case the drive's name
contains the default separator. Be carefull not to use '?' or '*' which could
exists in the reply of hddtemp.


CONTACT
=======
Emmanuel Varagnat: hddtemp@guzu.net

http://www.guzu.net


debian/changelog
hddtemp (0.3-beta15-53) unstable; urgency=medium

  * src/backtrace.c: rename "struct ucontext" into "struct ucontext_t" to
    fix FTBFS with glibc 2.26 (closes: bug#887753).
  * debian/control: change priority to optional.
  * debian/control: drop Suggests: ksensors.
  * debian/control: bumped Standards-Version to 4.1.3.
  * debian/rules: rewrite using dh.
  * Added Brazilian Portuguese debconf templates. Thanks to Diego Neves
    (closes: bug#836424).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 21 Jan 2018 14:19:42 +0100

hddtemp (0.3-beta15-52) unstable; urgency=low

  * Decrease hddtemp/daemon priority to medium from high (closes: 
    bug#680877).
  * Bumped Standards-Version to 3.9.3.

 -- Aurelien Jarno <aurel32@debian.org>  Mon, 09 Jul 2012 08:12:30 +0200

hddtemp (0.3-beta15-51) unstable; urgency=low

  * init script: also probe for double letter drives (closes: bug#646724).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 30 Oct 2011 20:21:06 +0100

hddtemp (0.3-beta15-50) unstable; urgency=low

  * Convert hddtemp.db to UTF-8 (closes: bug#640899).
  * Don't restart on reload in init.d (closes: bug#643610).
  * Add a status option in init.d (closes: bug#643611).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 09 Oct 2011 20:16:06 +0200

hddtemp (0.3-beta15-49) unstable; urgency=low

  * Added Slovak debconf templates. Thanks to Slavko (closes: bug#634519).
  * Fix a few typo in translation and enable German translation. Thanks to
    Ville Skyttä for tha patch.
  * Add build-arch and build-indep targets.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 30 Aug 2011 16:32:45 +0200

hddtemp (0.3-beta15-48) unstable; urgency=low

  * Fix backtrace support on i386.
  * Add backtrace support on amd64.

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 05 Jun 2011 00:18:46 +0200

hddtemp (0.3-beta15-47) unstable; urgency=low

  * Fix reading model on SATA big-endian (closes: #602307).
  * Updated Arabic debconf templates, thanks to Ossama Khaya (closes: 
    bug#596177).
  * Updated Swedish debconf template, thanks to Joe Hansen (closes:
    bug#603109).
  * Update Dutch debconf translation, thanks to Jeroen Schot (closes: 
    bug#628757).
  * Bumped Standards-Version to 3.9.2 (no changes).

 -- Aurelien Jarno <aurel32@debian.org>  Sat, 04 Jun 2011 21:36:32 +0200

hddtemp (0.3-beta15-46) unstable; urgency=low

  * Don't probe /dev/sg? by default (closes: bug#531849).
  * Add initial German translation, by Helge Kreutzmann (closes: #503710).
  * Fix a typo in the help message (closes: #503711).
  * Correctly zero-terminate long product ID strings (closes: #517981).
  * Bumped Standards-Version to 3.8.4 (no changes).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 07 Feb 2010 20:24:49 +0100

hddtemp (0.3-beta15-45) unstable; urgency=low

  * Updated Swedish debconf templates. Thanks to Martin Bagge (closes: 
    bug#491768).
  * Improve init script, by Cristian Ionescu-Idbohrn (closes: bug#486235).
  * Mention /etc/init.d/hddtemp and /etc/default/hddtemp in README.Debian
    (closes: bug#482982).
  * Bumped Standards-Version to 3.8.0 (no changes).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 27 Jul 2008 09:44:06 +0200

hddtemp (0.3-beta15-44) unstable; urgency=low

  * Updated Japanese debconf templates. Thanks to Kenshi Muto (closes:
    bug#483504).

 -- Aurelien Jarno <aurel32@debian.org>  Mon, 02 Jun 2008 10:30:58 +0200

hddtemp (0.3-beta15-43) unstable; urgency=low

  * Don't wake up SATA drives if not asked (closes: #479840). 

 -- Aurelien Jarno <aurel32@debian.org>  Sat, 10 May 2008 13:24:18 +0200

hddtemp (0.3-beta15-42) unstable; urgency=low

  * Use minus instead of hyphenin manpage. 

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 02 Mar 2008 19:14:14 +0100

hddtemp (0.3-beta15-41) unstable; urgency=low

  * Fixed LSB header in init script (closes: bug#466281).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 17 Feb 2008 18:22:21 +0100

hddtemp (0.3-beta15-40) unstable; urgency=low

  * Added Basque debconf templates. Thanks to Piarres Beobide (closes:
    bug#465731).

 -- Aurelien Jarno <aurel32@debian.org>  Fri, 15 Feb 2008 21:59:07 +0100

hddtemp (0.3-beta15-39) unstable; urgency=low

  [ Aurelien Jarno ]
  * Bumped Standards-Version to 3.7.3 (no changes).

  [ Christian Perrier ]
  * Debconf templates and debian/control reviewed by the debian-l10n-
    english team as part of the Smith review project. Closes: #462483
  * [Debconf translation updates]
  * Norwegian Bokmål. Closes: #462799
  * German. Closes: #462843
  * Galician. Closes: #462988
  * Spanish. Closes: #463898
  * Finnish. Closes: #463943
  * Czech. Closes: #464224
  * Russian. Closes: #464566
  * Italian. Closes: #465069
  * Portuguese. Closes: #465135
  * Vietnamese. Closes: #465534
  * French. Closes: #465650

 -- Aurelien Jarno <aurel32@debian.org>  Wed, 13 Feb 2008 23:43:53 +0100

hddtemp (0.3-beta15-38) unstable; urgency=low

  * Fix typo in hddtemp manpage (closes: bug#444242).

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 11 Dec 2007 15:17:16 +0100

hddtemp (0.3-beta15-37) unstable; urgency=low

  * Updated Czech debconf templates. Thanks to Miroslav Kure (closes:
    bug#413322).
  * Added Galician debconf templates. Thanks to Jacobo Tarrio (closes:
    bug#412198).
  * Clean po/stamp-po to make sure the .gmo files are regenerated (closes:
    bug#441708).

src/ata.c
/*
 * Copyright (C) 2002  Emmanuel VARAGNAT <hddtemp@guzu.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */


// Include file generated by ./configure
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

// Gettext includes
#if ENABLE_NLS
#include <libintl.h>
#define _(String) gettext (String)
#else
#define _(String) (String)
#endif

// Standard includes
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <linux/hdreg.h>

// Application specific includes
#include "hddtemp.h"
#include "atacmds.h"


#define SBUFF_SIZE 512
#define swapb(x) \
({ \
        u16 __x = (x); \
        (x) = ((u16)( \
                (((u16)(__x) & (u16)0x00ffU) << 8) | \
                (((u16)(__x) & (u16)0xff00U) >> 8) )); \
})

static char sbuff[SBUFF_SIZE];

static int ata_probe(int device) {
  if(device == -1 || ioctl(device, HDIO_GET_IDENTITY, sbuff))
    return 0;
  else
    return 1;
}

static struct disk_info ata_info (int device) {
  struct disk_info info;
  info.serial_number = strdup(_("unknown"));

  if(device == -1 || ioctl(device, HDIO_GET_IDENTITY, sbuff))
    info.model = strdup(_("unknown"));
  else
    info.model = strdup((char*) ((u16*)sbuff + 27));

  return info;
}


static enum e_gettemp ata_get_temperature(struct disk *dsk) {   
  unsigned char    values[512]/*, thresholds[512]*/;
  unsigned char *  field;
  int              i;
  u16 *            p;

  if(dsk->db_entry->attribute_id == 0) {
    close(dsk->fd);
    dsk->fd = -1;
    return GETTEMP_NOSENSOR;
  }

  if(ata_get_packet(dsk->fd)) {
    snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, _("S.M.A.R.T. not available"));
    return GETTEMP_NOT_APPLICABLE;
  }

  switch(ata_get_powermode(dsk->fd)) {
  case PWM_STANDBY:
  case PWM_SLEEPING:
    if (!wakeup)
      return GETTEMP_DRIVE_SLEEP;
  case PWM_UNKNOWN:
  case PWM_ACTIVE: /* active or idle */
  default:
    break;
  }

  /* get SMART values */
  if(ata_enable_smart(dsk->fd) != 0) {
    enum e_gettemp ret;
    if(errno == EIO) {
      snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, _("S.M.A.R.T. not available"));
      ret = GETTEMP_NOT_APPLICABLE;
    }
    else
    {
      snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, "%s", strerror(errno));
      ret = GETTEMP_ERROR;
    }
    close(dsk->fd);
    dsk->fd = -1;
    return ret;
  }

  if(ata_get_smart_values(dsk->fd, values)) {
    snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, "%s", strerror(errno));
    close(dsk->fd);
    dsk->fd = -1;
    return GETTEMP_ERROR;
  }

  p = (u16*)values;
  for(i = 0; i < 256; i++) {
    swapb(*(p+i));
  }

  /* get SMART threshold values */
  /*
  if(get_smart_threshold_values(fd, thresholds)) {
    perror("ioctl");
    exit(3);
  }

  p = (u16*)thresholds;
  for(i = 0; i < 256; i++) {
    swapb(*(p+i));
  }
  */

  /* temperature */
  field = ata_search_temperature(values, dsk->db_entry->attribute_id);

  if(field)
    dsk->value = *(field+3);

  if(dsk->value != -1)
    return GETTEMP_KNOWN;
  else
    return GETTEMP_UNKNOWN;

  /* never reached */
}


/*******************************
 *******************************/

struct bustype ata_bus = {
  "PATA",
  ata_probe,
  ata_info,
  ata_get_temperature
};


src/scsi.c
/*
 * Copyright (C) 2002  Emmanuel VARAGNAT <hddtemp@guzu.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
/* 
 * Adapted from a patch sended by : Frederic LOCHON <lochon@roulaise.net>
 */

// Include file generated by ./configure
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

// Gettext includes
#if ENABLE_NLS
#include <libintl.h>
#define _(String) gettext (String)
#else
#define _(String) (String)
#endif

// Standard includes
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <scsi/scsi.h>

// Application specific includes
#include "scsicmds.h"
#include "hddtemp.h"

static int scsi_probe(int device) {
  int bus_num;

  if(ioctl(device, SCSI_IOCTL_GET_BUS_NUMBER, &bus_num))
    return 0;
  else
    return 1;
}

static struct disk_info scsi_info (int device) {
  unsigned char buf[36];
  struct disk_info info;
  info.serial_number = strdup(_("unknown"));

  if (scsi_inquiry(device, buf) != 0) {
    info.model = strdup(_("unknown"));
  }
  else {
    info.model = strdup(buf + 8);
  }

  return info;
}

static enum e_gettemp scsi_get_temperature(struct disk *dsk) {
  int              i;
  int              tempPage = 0;
  unsigned char    buffer[1024];

  /*
    on triche un peu
  */
  dsk->db_entry = (struct harddrive_entry*) malloc(sizeof(struct harddrive_entry));
  if(dsk->db_entry == NULL) {
    perror("malloc");
    exit(-1);
  }
  
  dsk->db_entry->regexp       = "";
  dsk->db_entry->description  = "";
  dsk->db_entry->attribute_id = 0;
  dsk->db_entry->unit         = 'C';
  dsk->db_entry->next         = NULL;

  if (scsi_smartsupport(dsk->fd) == 0) {
    snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, _("S.M.A.R.T. not available"));
    close(dsk->fd);
    dsk->fd = -1;
    return GETTEMP_NOT_APPLICABLE;
  }

  /*
    Enable SMART
  */
  if (scsi_smartDEXCPTdisable(dsk->fd) != 0) {
    snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, "%s", strerror(errno));
    close(dsk->fd);
    dsk->fd = -1;
    return GETTEMP_ERROR;
  }

  /*
    Temp. capable 
  */  
  if (scsi_logsense(dsk->fd , SUPPORT_LOG_PAGES, buffer, sizeof(buffer)) != 0) {
    snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, _("log sense failed : %s"), strerror(errno));
    close(dsk->fd);
    dsk->fd = -1;
    return GETTEMP_ERROR;
   }

   for ( i = 4; i < buffer[3] + LOGPAGEHDRSIZE ; i++) {
     if (buffer[i] == TEMPERATURE_PAGE) {
       tempPage = 1;
       break;
     }
   }

   if(tempPage) {
      /* 
	 get temperature (from scsiGetTemp (scsicmd.c))
      */
      if (scsi_logsense(dsk->fd , TEMPERATURE_PAGE, buffer, sizeof(buffer)) != 0) {
	snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, _("log sense failed : %s"), strerror(errno));
	close(dsk->fd);
	dsk->fd = -1;
	return GETTEMP_ERROR;
      }

      dsk->value = buffer[9];

      return GETTEMP_KNOWN;
   } else {
     return GETTEMP_NOSENSOR;
   }
}

/*******************************
 *******************************/

struct bustype scsi_bus = {
  "SCSI",
  scsi_probe,
  scsi_info,
  scsi_get_temperature
};


src/hddtemp.c
/*
 * Copyright (C) 2002  Emmanuel VARAGNAT <hddtemp@guzu.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/*
 * Added some SCSI support: Frederic LOCHON <lochon@roulaise.net>
 *
 */

// Include file generated by ./configure
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

// Gettext includes
#if ENABLE_NLS
#include <libintl.h>
#define _(String) gettext (String)
#else
#define _(String) (String)
#endif

// Standard includes
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <getopt.h>
#include <signal.h>
#include <errno.h>
#include <locale.h>
#include <time.h>
#include <netinet/in.h>
#include <linux/hdreg.h>
#include <ctype.h>
#include <assert.h>

// Application specific includes
#include "ata.h"
#include "utf8.h"
#include "sata.h"
#include "scsi.h"
#include "db.h"
#include "hddtemp.h"
#include "backtrace.h"
#include "daemon.h"


#define PORT_NUMBER            7634
#define SEPARATOR              '|'
#define CACHE_DELAY            60

char *             database_path = DEFAULT_DATABASE_PATH;
long               portnum, syslog_interval;
char *             listen_addr;
char               separator = SEPARATOR;

struct bustype *   bus[BUS_TYPE_MAX];
int                tcp_daemon, debug, quiet, numeric, wakeup, foreground, af_hint, cache_delay;

static enum { DEFAULT, CELSIUS, FAHRENHEIT } unit;

/*******************************************************
 *******************************************************/

static void init_bus_types() {
  bus[BUS_SATA] = &sata_bus;
  bus[BUS_ATA] = &ata_bus;
  bus[BUS_SCSI] = &scsi_bus;
}

/*******************************************************
 *******************************************************/

int value_to_unit(struct disk *dsk) {
  switch(unit) {
  case CELSIUS:
    if(dsk->db_entry->unit == 'F')
      return F_to_C(dsk->value);
    break;
  case FAHRENHEIT:
    if(dsk->db_entry->unit == 'C')
      return C_to_F(dsk->value);
  default:
    break;
  }

  return dsk->value;
}

char get_unit(struct disk *dsk) {
  switch(unit) {
  case CELSIUS:
    return 'C';
  case FAHRENHEIT:
    return 'F';
  default:
    return dsk->db_entry->unit;
  }
}



static enum e_bustype probe_bus_type(struct disk *dsk) {
  /* SATA disks answer to both ATA and SCSI commands so
     they have to be probed first in order to be detected */
  if(bus[BUS_SATA]->probe(dsk->fd))
    return BUS_SATA;
  else if(bus[BUS_ATA]->probe(dsk->fd))
    return BUS_ATA;	  
  else if(bus[BUS_SCSI]->probe(dsk->fd))
    return BUS_SCSI;
  else
    return BUS_UNKNOWN;
}

/*
static int get_smart_threshold_values(int fd, unsigned char* buff) {
  unsigned char cmd[516] = { WIN_SMART, 0, SMART_READ_THRESHOLDS, 1 };
  int ret;

  ret = ioctl(fd, HDIO_DRIVE_CMD, cmd);
  if(ret)
    return ret;

  memcpy(buff, cmd+4, 512);
  return 0;
}
*/


static void display_temperature(struct disk *dsk) {
  enum e_gettemp ret;
  char *degree;

  if(dsk->type != ERROR && debug ) {
    printf(_("\n================= hddtemp %s ==================\n"
             "Model: %s Serial: %s\n\n"), VERSION, dsk->info.model, dsk->info.serial_number);
    /*    return;*/
  }

  if(dsk->type == ERROR
     || bus[dsk->type]->get_temperature == NULL
     || (ret = bus[dsk->type]->get_temperature(dsk)) == GETTEMP_ERROR )
  {
    fprintf(stderr, "%s: %s\n", dsk->drive, dsk->errormsg);
    return;
  }

  if(debug)
    return;

  degree = degree_sign();
  switch(ret) {
  case GETTEMP_ERROR:
    /* see above */
    break;
  case GETTEMP_NOT_APPLICABLE:

    if (numeric && quiet)
      printf("0\n");      
    else
      printf("%s: %s: %s: %s\n", dsk->drive, dsk->info.model, dsk->info.serial_number, dsk->errormsg);

    break;
  case GETTEMP_UNKNOWN:

    if(!quiet)
      fprintf(stderr,
	      _("WARNING: Drive %s doesn't seem to have a temperature sensor.\n"
		"WARNING: This doesn't mean it hasn't got one.\n"
		"WARNING: If you are sure it has one, please contact me (hddtemp@guzu.net).\n"
		"WARNING: See --help, --debug and --drivebase options.\n"), dsk->drive);

    if (numeric && quiet)
      printf("0\n");      
    else
      fprintf(stderr, _("%s: %s: %s: no sensor\n"), dsk->drive, dsk->info.model, dsk->info.serial_number);

    break;
  case GETTEMP_KNOWN:

    if (! numeric)
       printf("%s: %s: %s: %d%s%c\n",
              dsk->drive,
              dsk->info.model,
              dsk->info.serial_number,
	      value_to_unit(dsk),
              degree,
	      get_unit(dsk)
	      );
    else
       printf("%d\n", value_to_unit(dsk));

    break;
  case GETTEMP_DRIVE_SLEEP:

    if (numeric && quiet)
      printf("0\n");      
    else
      fprintf(stderr, _("%s: %s: %s: drive is sleeping\n"), dsk->drive, dsk->info.model, dsk->info.serial_number);

    break;
  case GETTEMP_NOSENSOR:
    if (numeric && quiet)
      printf("0\n");      
    else
      fprintf(stderr, _("%s: %s: %s: drive supported, but it doesn't have a temperature sensor.\n"), dsk->drive, dsk->info.model, dsk->info.serial_number);
      
    break;
  default:
    fprintf(stderr, _("ERROR: %s: %s: %s: unknown returned status\n"), dsk->drive, dsk->info.model, dsk->info.serial_number);
    break;
  }
  free(degree);
}


void do_direct_mode(struct disk *ldisks) {
  struct disk *dsk;

  for(dsk = ldisks; dsk; dsk = dsk->next) {
    display_temperature(dsk);
  }
  
  if(debug) {
    printf(_("\n"
	     "If one of the field value seems to match the temperature, be sure to read\n"
	     "the hddtemp man page before sending a report (section REPORT). Thanks.\n"
	     ));
  }
}


int main(int argc, char* argv[]) {
  int           i, c, lindex = 0, db_loaded = 0;
  int 		ret = 0;
  int           show_db;
  struct        disk * ldisks;

  backtrace_sigsegv();
  backtrace_sigill();
  backtrace_sigbus();

  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
  
  show_db = debug = numeric = quiet = wakeup = af_hint = syslog_interval = foreground = 0;
  unit = DEFAULT;
  portnum = PORT_NUMBER;
  cache_delay = CACHE_DELAY;
  listen_addr = NULL;

  /* Parse command line */
  optind = 0;
  while (1) {
    static struct option long_options[] = {
      {"help",       0, NULL, 'h'},
      {"quiet",      0, NULL, 'q'},
      {"daemon",     0, NULL, 'd'},
      {"drivebase",  0, NULL, 'b'},
      {"debug",      0, NULL, 'D'},
      {"file",       1, NULL, 'f'},
      {"foreground", 0, NULL, 'F'},
      {"listen",     1, NULL, 'l'},
      {"version",    0, NULL, 'v'},
      {"port",       1, NULL, 'p'},
      {"separator",  1, NULL, 's'},
      {"numeric",    0, NULL, 'n'},
      {"unit",       1, NULL, 'u'},
      {"syslog",     1, NULL, 'S'},
      {"wake-up",    0, NULL, 'w'},
      {"cache-delay",1, NULL, 'c'},
      {0, 0, 0, 0}
    };
 
    c = getopt_long (argc, argv, "bDdf:l:hp:c:qs:u:vnw46FS:", long_options, &lindex);
    if (c == -1)
      break;
    
    switch (c) {
      case 'q':
	quiet = 1;
	break;
      case '4':
	af_hint = AF_INET;
	break;
      case '6':
	af_hint = AF_INET6;
	break;
      case 'b':
	show_db = 1;
	break;
      case 'd':
	tcp_daemon = 1;
	break;
      case 'D':
	debug = 1;
	break;
      case 'f':
	database_path = optarg;
	break;
      case 's':
	separator = optarg[0];
	if(separator == '\0') {
	  fprintf(stderr, _("ERROR: invalid separator.\n"));
	  exit(1);
	}
	break;
      case 'p':
	{
	  char *end = NULL;

	  portnum = strtol(optarg, &end, 10);

	  if(errno == ERANGE || end == optarg || *end != '\0' || portnum < 1) {
	    fprintf(stderr, _("ERROR: invalid port number.\n"));
	    exit(1);
	  }
	}
	break;
      case 'c':
	{
	  char *end = NULL;

	  cache_delay = strtol(optarg, &end, 10);

	  if(errno == ERANGE || end == optarg || *end != '\0' || cache_delay < 2) {
	    fprintf(stderr, _("ERROR: invalid cache delay.\n"));
	    exit(1);
	  }
	}
	break;
      case 'u':
	switch(*optarg) {
	case 'c':
	case 'C':
	  unit = CELSIUS;
	  break;
	case 'f':
	case 'F':
	  unit = FAHRENHEIT;
	  break;
	}
	break;
      case 'l':
	listen_addr = optarg;
	break;
      case '?':
      case 'h':
	printf(_(" Usage: hddtemp [OPTIONS] [TYPE:]DISK1 [[TYPE:]DISK2]...\n"
		 "\n"
		 "   hddtemp displays the temperature of drives supplied in argument.\n"
		 "   Drives must support S.M.A.R.T.\n"
		 "\n"
		 "  TYPE could be SATA, PATA or SCSI. If omitted hddtemp will try to guess.\n"
		 "\n"
		 "  -b   --drivebase     :  display database file content that allow hddtemp to\n"
		 "                          recognize supported drives.\n"
		 "  -D   --debug         :  display various S.M.A.R.T. fields and their values.\n"
		 "                          Useful to find a value that seems to match the\n"
		 "                          temperature and/or to send me a report.\n"
		 "                          (done for every drive supplied).\n"
		 "  -d   --daemon        :  run hddtemp in TCP/IP daemon mode (port %d by default.)\n"
		 "  -f   --file=FILE     :  specify database file to use.\n"
		 "  -F   --foreground    :  don't daemonize, stay in foreground.\n"
		 "  -l   --listen=addr   :  listen on a specific interface (in TCP/IP daemon mode).\n"
		 "  -n   --numeric       :  print only the temperature.\n"
		 "  -p   --port=#        :  port to listen to (in TCP/IP daemon mode).\n"
		 "  -s   --separator=C   :  separator to use between fields (in TCP/IP daemon mode).\n"
		 "  -S   --syslog=s      :  log temperature to syslog every s seconds.\n"
		 "  -u   --unit=[C|F]    :  force output temperature either in Celsius or Fahrenheit.\n"
		 "  -q   --quiet         :  do not check if the drive is supported.\n"
		 "  -v   --version       :  display hddtemp version number.\n"
		 "  -w   --wake-up       :  wake-up the drive if need.\n"
		 "  -4                   :  listen on IPv4 sockets only.\n"
		 "  -6                   :  listen on IPv6 sockets only.\n"
		 "  -c   --cache-delay=# :  how long the daemon caches tempeature values for, in\n"
		 "                          seconds. The minimium value is 2. The default is %d.\n"
		 "\n"
		 "Report bugs or new drives to <hddtemp@guzu.net>.\n"),
	       PORT_NUMBER, CACHE_DELAY);
	break; 
      case 'v':
	printf(_("hddtemp version %s\n"), VERSION);
	exit(0);
	break; 
      case 'n':
        numeric = 1;
        break;
      case 'w':
	wakeup = 1;
	break;
      case 'S':
	{
	  char *end = NULL;

	  syslog_interval = strtol(optarg, &end, 10);

	  if(errno == ERANGE || end == optarg || *end != '\0' || syslog_interval < 1) {
	    fprintf(stderr, _("ERROR: invalid interval.\n"));
	    exit(1);
	  }
        }
	break;
      case 'F':
        foreground = 1;
	break;
      default:
	exit(1);
      }
  }
  
  if(show_db) {
     load_database(database_path);
     display_supported_drives();
     exit(0);
  }
  
  if(argc - optind <= 0) {
    fprintf(stderr, _("Too few arguments: you must specify one drive, at least.\n"));
    exit(1);
  }

  if(debug) {
    /*    argc = optind + 1;*/
    quiet = 1;
  }

  if(debug && (tcp_daemon || syslog_interval != 0)) {
    fprintf(stderr, _("ERROR: can't use --debug and --daemon or --syslog options together.\n"));
    exit(1);
  }

  init_bus_types();

  /* collect disks informations */
  ldisks = NULL;
  for(i = argc - 1; i >= optind; i--) {
    struct disk *dsk = (struct disk *) malloc(sizeof(struct disk));
    char *p;

    assert(dsk);

    memset(dsk, 0, sizeof(*dsk));

    p = strchr(argv[i], ':');
    if(p == NULL)
      dsk->drive = argv[i];
    else {
      char *q;
      int j;

      /* upper case type */
      for(q = argv[i]; q != p; q++)
	*q = (char) toupper(*q);

      /* force bus type */
      for(j = 0; j < BUS_TYPE_MAX; j++) {
	if(bus[j] &&
	   bus[j]->name &&
	   strncmp(bus[j]->name, argv[i], p - argv[i] - 1) == 0)
	  {
	    dsk->type = j;
	    break;
	  }
      }

      dsk->drive = p + 1;      
    }

    dsk->next = ldisks;
    ldisks = dsk;    

    errno = 0;
    dsk->errormsg[0] = '\0';
    if( (dsk->fd = open(dsk->drive, O_RDONLY | O_NONBLOCK)) < 0) {
      snprintf(dsk->errormsg, MAX_ERRORMSG_SIZE, "open: %s\n", strerror(errno));
      dsk->type = ERROR;
      ret = 1;
      continue;
    }

    if( ! dsk->type ) {
      dsk->type = probe_bus_type(dsk);
    }

    if(dsk->type == BUS_UNKNOWN) {
      fprintf(stderr, _("ERROR: %s: can't determine bus type (or this bus type is unknown)\n"), dsk->drive);

      ldisks = dsk->next;
      free(dsk);
      ret = 1;
      continue;
    }

    dsk->info = bus[dsk->type]->info(dsk->fd);
    dsk->value = -1;
    if(dsk->type != BUS_SCSI) {
      struct harddrive_entry   *dbe;

      if(!db_loaded) {
	load_database(database_path);
	db_loaded = 1;
      }      

      dsk->db_entry = (struct harddrive_entry *)malloc(sizeof(struct harddrive_entry));
      dbe = is_a_supported_drive(dsk->info.model);
      if(dbe)
	memcpy(dsk->db_entry, dbe, sizeof(struct harddrive_entry));
      else {
        dsk->db_entry->regexp       = "";
        dsk->db_entry->description  = "";
        dsk->db_entry->attribute_id = DEFAULT_ATTRIBUTE_ID;
        dsk->db_entry->unit         = 'C';
        dsk->db_entry->next         = NULL;
      }
    }
  }

  free_database();
  if(tcp_daemon || syslog_interval != 0) {
    do_daemon_mode(ldisks);
  }
  else {
    do_direct_mode(ldisks);
  }

  return ret;
}
